### ISSUE TITLE
[Bug]: WITH (UPDLOCK) does not create a proper row lock

### ISSUE THREAD

**Issue by wieck** [babelfish-for-postgresql/babelfish_extensions#88]
### Contact Details

jan@wi3ck.info

### What happened?

This happens with the [BenchmarkSQL](https://github.com/wieck/benchmarksql), which is a TPC-C style benchmark. The exact same code works fine with a real MS-SQL Server 2017.

The (relevant) queries of the NEW_ORDER transaction are

```
SELECT d_tax, d_next_o_id
    FROM bmsql_district WITH (UPDLOCK)
    WHERE d_w_id = ? AND d_id = ?

UPDATE bmsql_district
    SET d_next_o_id = d_next_o_id + 1
    WHERE d_w_id = ? AND d_id = ?

INSERT INTO bmsql_oorder (
    o_id, o_d_id, o_w_id, o_c_id, o_entry_d,
    o_ol_cnt, o_all_local) 
    VALUES (?, ?, ?, ?, ?, ?, ?)
```
These are all JDBC PreparedStatement constructs. But that should not make any difference.

**WITH (UPDLOCK)** is T-SQL's equivalent of the SQL standard **FOR UPDATE**. Babelfish seems to accept the syntax but with no effect on the actual locking behavior. Under enough concurrency there are instances of transactions getting aborted due to

```
2022-03-12 19:32:41.154 UTC [56660] ERROR:  duplicate key value violates unique constraint "bmsql_oorder_pkeybmsql_oorder3723443fbaae7a4abcf8aec1f6574105"
2022-03-12 19:32:41.154 UTC [56660] DETAIL:  Key (o_w_id, o_d_id, o_id)=(8, 7, 3088) already exists.
2022-03-12 19:32:41.154 UTC [56660] CONTEXT:  SQL statement "INSERT INTO bmsql_oorder (    o_id, o_d_id, o_w_id, o_c_id, o_entry_d,     o_ol_cnt, o_all_local) VALUES ("@P0", "@P1", "@P2", "@P3", "@P4", "@P5", "@P6")"
	PL/tsql function inline_code_block line 1 at SQL statement
	TDS Protocol: Message Type: RPC, SP Type: SP_EXECUTE, Phase: TDS_REQUEST_PHASE_PROCESS. Processing SP_EXECUTE Request
```

This indicates that more than one session is able to retrieve the same row version from bmsql_district **WITH UPDLOCK** before the UPDATE happens, which should not be possible.

### Version

BABEL_1_X_DEV (Default)

### Extension

babelfishpg_tsql (Default)

### What flavour of Linux are you running into the bug?

Other

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by wieck** [comment]
I can also reproduce the problem with two manual SQLCMD sessions. Babelfish indeed does allow two sessions to perform

```
BEGIN TRAN
GO
SELECT ... WITH (UPDLOCK) ...
GO
```
for the exact same row in two sessions. And both sessions do retrieve the same row version. Even if the first session does an UPDATE, the second is not blocked from doing the SELECT (and getting the old row version because the UPDATE hasn't been committed yet).

This is the PostgreSQL behavior of SELECT and UPDATE **without** using FOR UPDATE. 

In other words, the `WITH (UPDLOCK)` syntax is only parsed and then thrown away. The required effect on locks does not happen.

**Comment by robverschoor** [comment]
This is a known current limitation. As documented, all hints are currently ignored.

**Comment by wieck** [comment]
This is not really a hint but a request to acquire a row level lock for the selected data. T-SQL does not support the ANSI-SQL `SELECT ... FOR UPDATE [NOWAIT]` syntax. That makes the `WITH (UPDLOCK)` a necessary, basic feature.

Is there a suggested workaround for this?

**Comment by robverschoor** [comment]
I do not disagree on the semantics, but technically this is a 'table hint' in the T-SQL dialect. As for a workaround, the same effect can be achieved by doing a dummy update first.

**Comment by wieck** [comment]
Thanks Rob.

I will try that workaround for now. It will double the number of dead rows for those actions (and the TPC-C has 3 of them), so it may have a negative impact on performance.

**Comment by robverschoor** [comment]
You can also generate a key value this way by first performing the update (SET d_next_o_id += 1) and then use the SELECT to pick up the generated value. This seems a little more direct than first doing a SELECT with an update lock and then doing the update.

**Comment by wieck** [comment]
It would work even better if T-SQL had support for `UPDATE ... RETURNING`.

For now I went with the dummy update first and can run BenchmarkSQL without errors. See [this commit](https://github.com/wieck/benchmarksql/commit/59998b497d4ab049a04c80d4ceaf2009c413af3a). I may change it later and see if it makes a performance difference.

I am currently trying to put a fix into the index build OOM problem so that I can compare Babelfish against plain PostgreSQL.

Thanks again.

**Comment by robverschoor** [comment]
You can do it as follows in SQL Server, but Babelfish does not currently support this syntax:

DECLARE @v int
UPDATE bmsql_district
    SET d_next_o_id = d_next_o_id + 1
                 @v = d_next_o_id + 1
WHERE [...]
SELECT @v    ==>  contains the generated key value

**Comment by wieck** [comment]
The application needs other column values from the DISTRICT, which are not updated, also. That is why doing a `SELECT ... FOR UPDATE` of all the needed information, then doing an `UPDATE ...` for just the one, single column that changes, is the cleanest and most easy to understand and maintain code. All other databases, Oracle, MS-SQL, PostgreSQL, MySQL and MariaDB support this. And this is not the only place where the TPC-C requires this access pattern. It actually has 3 places where `... FOR UPDATE` is needed.

I would put this feature rather high on the priority list.

**Comment by robverschoor** [comment]
The purpose of the UPDLOCK is to serialize on the key generation. The additional attributes can be retrieved after that. There are multiple ways of doing this though.

**Comment by wieck** [comment]
Rob, one of the primary goals for Babelfish was to provide a migration path for legacy applications without the need of invasive code changes. Rewriting a simple `SELECT ... WITH (UPDLOCK)` followed by an `UPDATE` into a completely reversed query logic is the exact opposite of that original goal. Especially when the correctness of such change can only be tested under concurrent load.

In this context I consider BenchmarkSQL an example of a legacy application. It works perfectly correct on MS-SQL 2017. Trying to run it on Babelfish requires to change the query logic in multiple places. The problem isn't even raised prominently, since the change in behavior is silent. One only runs into duplicate key errors when running this under high enough concurrency.

There is another place in the TPC-C where the UPDLOCK is essential. That is the PAYMENT transaction for customers with bad credit. It has to have a lock on the CUSTOMER row when retrieving the credit status to determine if it actually needs the rather large and network expensive c_data column. And only when that happens it performs rather complicated string manipulation on that data and finally uses a different query to update the CUSTOMER row by providing the new, client application generated, data for that column. 

So while I agree with you that there are many ways to work around this missing feature on a case by case base, it is a bug infested mine field and needs attention ASAP.

**Comment by robverschoor** [comment]
I do not disagree that this is a feature that may be used by existing applications and which Babelfish should ideally support. But I thought we were discussing possible workarounds here, hence my comments.

**Comment by suprio-amzn** [comment]
Table Hints are currently not supported. The behavior to give errors is controlled by the escape hatch `babelfishpg_tsql.escape_hatch_table_hints`.
